src/target/arc32.c

```diff
@@ -418,36 +420,82 @@ int arc32_config_step(struct target *target, int enable_step)
 	return ERROR_OK;
 }
 
-/* This function is cheap to call and returns quickly if caches already has
- * been invalidated since core had been halted. */
-int arc32_cache_invalidate(struct target *target)
+static int arc32_icache_invalidate(struct arc32_common *arc32)
 {
-	uint32_t value, backup;
+	uint32_t value = IC_IVIC_INVALIDATE;	/* invalidate I$ */
 
-	struct arc32_common *arc32 = target_to_arc32(target);
-
-	/* Don't waste time if already done. */
-	if (arc32->cache_invalidated)
+	if (arc32->icache_invalidated)
 	    return ERROR_OK;
 
-	LOG_DEBUG("Invalidating I$ & D$.");
+	LOG_DEBUG("Invalidating I$.");
 
-	value = IC_IVIC_INVALIDATE;	/* invalidate I$ */
 	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_IC_IVIC_REG, value));
-	CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, &value));
+	arc32->icache_invalidated = true;
+
+	return ERROR_OK;
+}
+
+static int arc32_dcache_invalidate(struct arc32_common *arc32)
+{
+	uint32_t value, backup;
+
+	if (arc32->dcache_invalidated)
+	    return ERROR_OK;
+
+	LOG_DEBUG("Invalidating D$.");
 
+	CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, &value));
 	backup = value;
-	value = value & ~DC_CTRL_IM;
+	value &= ~DC_CTRL_IM;
 
 	/* set DC_CTRL invalidate mode to invalidate-only (no flushing!!) */
 	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, value));
 	value = DC_IVDC_INVALIDATE;	/* invalidate D$ */
 	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_IVDC_REG, value));
-
 	/* restore DC_CTRL invalidate mode */
 	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, backup));
+	arc32->dcache_invalidated = true;
+
+	return ERROR_OK;
+}
+
+static int arc32_l2cache_invalidate(struct arc32_common *arc32)
+{
+	uint32_t value, backup;
+
+	if (!arc32->has_l2cache || arc32->l2cache_invalidated)
+	    return ERROR_OK;
+
+	LOG_DEBUG("Invalidating L2$.");
+
+	CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_L2_CTRL_REG, &value));
+	backup = value;
+	value &= ~L2_CTRL_IM;
+
+	/* set L2_CTRL invalidate mode to invalidate-only (no flushing!!) */
+	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_L2_CTRL_REG, value));
+	value = L2_IVL2_INVALIDATE; /* invalidate L2$ */
+	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_L2_IVL2_REG, value));
+	/* Wait until invalidate operation ends */
+	do {
+	    CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_L2_CTRL_REG, &value));
+	} while (value & L2_CTRL_BS);
+	/* restore L2_CTRL invalidate mode */
+	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_L2_CTRL_REG, backup));
+	arc32->l2cache_invalidated = true;
 
-	arc32->cache_invalidated = true;
+	return ERROR_OK;
+}
+
+
+/* This function is cheap to call and returns quickly if caches already has
+ * been invalidated since core had been halted. */
```

Comment:
This comment is misplaced, because *this* function itself doesn't return quickly - it is functions that it calls can return quickly`.